private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -386,31 +371,92 @@ private static String toHexS(String fmt, String algorithm, byte[] toHash) {
   }
 
   /**
-   * Attempts to download the file at the given targetUrl.
-   * Valid protocols are: http(s) & file at the moment.
+   * Attempts to download the file at the given targetUrl. Valid protocols are: http(s) & file at
+   * the moment.
    */
   private static CompletableFuture<Optional<String>> download(
-          String targetUrl, Credentials credentials
-  ) {
+      String targetUrl, Credentials credentials) {
     if (targetUrl.startsWith("http")) {
-      return CompletableFuture.supplyAsync( () -> {
+      return httpDownload(targetUrl, credentials);
+    } else if (targetUrl.startsWith("file://")) {
+      return fileDownload(targetUrl);
+    } else if (targetUrl.startsWith("s3://")) {
+      return s3Download(targetUrl);
+    } else {
+      throw new IllegalArgumentException("Unsupported protocol for download.");
+    }
+  }
+
+  private static CompletableFuture<Optional<String>> s3Download(String targetUrl) {
+    return CompletableFuture.supplyAsync(
+        () -> {
+          S3Client s3Client = S3Client.create();
+          try {
+            URI s3Uri = new URI(targetUrl);
+            String bucketName = s3Uri.getHost();
+            String path = s3Uri.getPath().substring(1);
```

Comment:
nit: i would `path` -> `key` better naming for S3.
I would break out this parsing maybe into URI -> S3Object